File: Golang/printListFromTailToHead.go
package main

import (
	"fmt"
)

type ListNode struct {
	data int
	succ *ListNode
}

func (list *ListNode) printListFromTailToHead() {
	if list == nil {
		return
	}
	var new_head *ListNode
	for list != nil {
		node := list
		list = list.succ
		new_head = insert(new_head, node)
	}
	for new_head != nil {
		fmt.Printf("%d ", new_head.data)
		new_head = new_head.succ
	}
}

func insert(head *ListNode, node *ListNode) *ListNode {
	node.succ = head
	return node
}

func main() {
	var head *ListNode
	for i := 0; i < 5; i++ {
		node := new(ListNode)
		node.data = i
		head = insert(head, node)
	}
	node := head
	for node != nil {
		fmt.Printf("%d ", node.data)
		node = node.succ
	}
	println()
	head.printListFromTailToHead()
	println()
}

File: Cplusplus/printListFromTailToHead.cpp
/**
*  struct ListNode {
*        int val;
*        struct ListNode *next;
*        ListNode(int x) :
*              val(x), next(NULL) {
*        }
*  };
*/
class Solution {
public:
    ListNode* insert(ListNode *head, ListNode *node) {
        node->next = head;
        return node;
    }
    vector<int> printListFromTailToHead(ListNode* head) {
        vector<int> vec;
        if (head == nullptr) {
            return vec;
        }
        ListNode *new_head = nullptr;
        while (head) {
            auto node = head;
            head = head->next;
            new_head = insert(new_head, node);
        }
        while(new_head != nullptr) {
            vec.push_back(new_head->val);
            new_head = new_head->next;
        }
        return vec;
    }
};